File: sdks/java/src/main/java/com/bcs/Uleb128.java
package com.bcs;

/**
 * ULEB128 encoding and decoding for BCS.
 *
 * <p>ULEB128 (Unsigned Little-Endian Base 128) is a variable-length encoding for unsigned integers.
 * Each byte contributes 7 bits of data, with the high bit indicating whether more bytes follow.</p>
 */
public final class Uleb128 {

  /** Maximum value that can be encoded (u32 max). */
  public static final long MAX_U32 = 0xFFFFFFFFL;

  private Uleb128() {
    // Utility class
  }

  /**
   * Encode an unsigned integer as ULEB128.
   *
   * @param value the value to encode (must be 0 <= value <= 2^32-1)
   * @return the ULEB128 encoded bytes
   * @throws IllegalArgumentException if value is negative or exceeds u32 max
   */
  public static byte[] encode(long value) {
    if (value < 0) {
      throw new IllegalArgumentException("ULEB128 cannot encode negative values: " + value);
    }
    if (value > MAX_U32) {
      throw new IllegalArgumentException("ULEB128 value exceeds u32 max: " + value);
    }

    // Fast path for small values (most common case)
    if (value < 0x80) {
      return new byte[] {(byte) value};
    }
    if (value < 0x4000) {
      return new byte[] {(byte) ((value & 0x7F) | 0x80), (byte) (value >>> 7)};
    }

    // General case: use fixed-size array (max 5 bytes for u32)
    byte[] buf = new byte[5];
    int len = 0;
    long remaining = value;

    do {
      int byteVal = (int) (remaining & 0x7F);
      remaining >>>= 7;
      if (remaining != 0) {
        byteVal |= 0x80;
      }
      buf[len++] = (byte) byteVal;
    } while (remaining != 0);

    // Return exact-sized array
    if (len == 5) {
      return buf;
    }
    byte[] result = new byte[len];
    System.arraycopy(buf, 0, result, 0, len);
    return result;
  }

  /**
   * Encode a ULEB128 value directly into a byte array.
   *
   * @param value the value to encode
   * @param dest the destination array
   * @param offset the offset to write at
   * @return the number of bytes written
   */
  public static int encodeTo(long value, byte[] dest, int offset) {
    if (value < 0 || value > MAX_U32) {
      throw new IllegalArgumentException("ULEB128 value out of range: " + value);
    }

    int pos = offset;
    long remaining = value;

    do {
      int byteVal = (int) (remaining & 0x7F);
      remaining >>>= 7;
      if (remaining != 0) {
        byteVal |= 0x80;
      }
      dest[pos++] = (byte) byteVal;
    } while (remaining != 0);

    return pos - offset;
  }

  /**
   * Decode a ULEB128 value from a byte array.
   *
   * @param data the data to decode from
   * @param offset the offset to start decoding from
   * @return a DecodeResult containing the decoded value and bytes consumed
   * @throws BcsError if decoding fails
   */
  public static DecodeResult decode(byte[] data, int offset) {
    long value = 0;
    int shift = 0;
    int bytesRead = 0;

    while (true) {
      if (offset + bytesRead >= data.length) {
        throw BcsError.unexpectedEof();
      }

      int byteVal = data[offset + bytesRead] & 0xFF;
      bytesRead++;

      // Check for overflow before adding (5 bytes max for u32)
      if (bytesRead == 5) {
        // Check for non-canonical encoding (could have been encoded in 4 bytes)
        if (byteVal == 0) {
          throw BcsError.nonCanonicalUleb128();
        }
        if (byteVal >= 0x10) {
          throw BcsError.uleb128Overflow();
        }
        // Final byte for 5-byte encoding
        value |= ((long) byteVal) << shift;
        if (value > MAX_U32) {
          throw BcsError.uleb128Overflow();
        }
        return new DecodeResult(value, bytesRead);
      }

      int digit = byteVal & 0x7F;
      value |= ((long) digit) << shift;

      if ((byteVal & 0x80) == 0) {
        // Last byte - check for non-canonical encoding
        if (bytesRead > 1 && byteVal == 0) {
          throw BcsError.nonCanonicalUleb128();
        }
        return new DecodeResult(value, bytesRead);
      }

      shift += 7;
    }
  }

  /**
   * Calculate the number of bytes needed to encode a value.
   *
   * @param value the value to measure
   * @return the number of bytes needed
   */
  public static int encodedSize(long value) {
    if (value == 0) {
      return 1;
    }

    int size = 0;
    long remaining = value;
    while (remaining > 0) {
      remaining >>>= 7;
      size++;
    }
    return size;
  }

  /** Result of ULEB128 decoding. */
  public static class DecodeResult {
    private final long value;
    private final int bytesConsumed;

    public DecodeResult(long value, int bytesConsumed) {
      this.value = value;
      this.bytesConsumed = bytesConsumed;
    }

    public long getValue() {
      return value;
    }

    public int getBytesConsumed() {
      return bytesConsumed;
    }
  }
}

File: sdks/kotlin/src/main/kotlin/com/bcs/Uleb128.kt
/**
 * ULEB128 encoding/decoding utilities
 */
package com.bcs

/**
 * ULEB128 utilities
 */
object Uleb128 {
    /** Maximum value that can be encoded as ULEB128 in BCS (u32 max) */
    const val MAX_VALUE: Long = 0xFFFFFFFFL

    /** Maximum number of bytes in a ULEB128-encoded u32 */
    const val MAX_BYTES: Int = 5

    /**
     * Encode a 32-bit unsigned integer as ULEB128
     */
    fun encode(value: Long): ByteArray {
        require(value >= 0 && value <= MAX_VALUE) { "Value out of range for ULEB128" }

        val result = mutableListOf<Byte>()
        var v = value

        do {
            var byte = (v and 0x7F).toInt()
            v = v shr 7
            if (v != 0L) {
                byte = byte or 0x80
            }
            result.add(byte.toByte())
        } while (v != 0L)

        return result.toByteArray()
    }

    /**
     * Encode a 32-bit unsigned integer as ULEB128
     */
    fun encode(value: Int): ByteArray = encode(value.toLong() and 0xFFFFFFFFL)

    /**
     * Decode a ULEB128-encoded value from bytes
     *
     * @return Pair of (decoded value, bytes consumed)
     */
    fun decode(data: ByteArray, offset: Int = 0): Pair<Long, Int> {
        var value = 0L
        var shift = 0
        var bytesRead = 0

        for (i in 0 until MAX_BYTES) {
            if (offset + i >= data.size) {
                throw BcsError.unexpectedEof()
            }

            val byte = data[offset + i].toInt() and 0xFF
            val digit = byte and 0x7F

            value = value or (digit.toLong() shl shift)
            bytesRead = i + 1

            // Check if this is the last byte (high bit not set)
            if ((byte and 0x80) == 0) {
                // Check for non-canonical encoding (trailing zeros)
                if (shift > 0 && digit == 0) {
                    throw BcsError.nonCanonicalUleb128()
                }

                // Check for overflow
                if (value > MAX_VALUE) {
                    throw BcsError.uleb128Overflow()
                }

                return Pair(value, bytesRead)
            }

            shift += 7
        }

        // If we've read MAX_BYTES and still have continuation bit, overflow
        throw BcsError.uleb128Overflow()
    }

    /**
     * Calculate the encoded size of a value
     */
    fun encodedSize(value: Long): Int {
        var size = 1
        var v = value
        while (v >= 0x80) {
            v = v shr 7
            size++
        }
        return size
    }
}
